### PR TITLE
Do bulk upsert in DocumentFetcher

### PR DESCRIPTION
**Pull request by yoomlam**
### Description
Refactors code to do batch UPDATEs, INSERTs, and SELECTs to reduce DB CPU load and be more scalable to large numbers of documents.

[Optimizes](https://dsva.slack.com/archives/CHD7QU4L8/p1608138158198700) `DocumentsController#index` queries to help address [PagerDuty alerts regarding High DB CPU usage](https://dsva.slack.com/archives/C4JECDLSE/p1607958440131700) and summarized in 3rd bullet in https://hackmd.io/SVO3cqLfQo2W1exQ2GUC9w?both#Next-Steps.

[NewRelic](https://one.newrelic.com/launcher/nr1-core.explorer?pane=eyJuZXJkbGV0SWQiOiJhcG0tbmVyZGxldHMudHJhbnNhY3Rpb25zIiwiZW50aXR5SWQiOiJNVGM0T0RRMU9IeEJVRTE4UVZCUVRFbERRVlJKVDA1OE9EQTJNRGcxTkRRIn0=&sidebars[0]=eyJuZXJkbGV0SWQiOiJucjEtY29yZS5hY3Rpb25zIiwiZW50aXR5SWQiOiJNVGM0T0RRMU9IeEJVRTE4UVZCUVRFbERRVlJKVDA1OE9EQTJNRGcxTkRRIiwic2VsZWN0ZWROZXJkbGV0Ijp7Im5lcmRsZXRJZCI6ImFwbS1uZXJkbGV0cy50cmFuc2FjdGlvbnMifX0=&platform[accountId]=1788458&platform[timeRange][duration]=1800000) shows `Reader::DocumentsController#index` to be one of the most time consuming queries:
image

Taking upwards of 35 seconds:
image

Potentially taking 30 seconds to complete:
image

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Reader fetches documents from eFolder and VBMS correctly and scalably

### Testing Plan
In master branch, open a Rails console and benchmark:
```ruby
last_doc_id = Document.last.id
# Create documents to be returned from document_fetcher
documents = Array.new(500) { Generators::Document.build }.uniq(&:vbms_document_id);

# Create some existing documents that match half of those in documents
saved_documents = Array.new(documents.size / 2) do |i|
            # have every other document already exists to test that all CREATEs and UPDATEs are each done once
            fetched_document = documents[i * 2]
            Generators::Document.create(
              type: "Form 9",
              series_id: fetched_document.series_id,
              vbms_document_id: fetched_document.vbms_document_id
            )
          end;

service_manifest_vbms_fetched_at = Time.zone.local(1989, "nov", 23, 8, 2, 55)
service_manifest_vva_fetched_at = Time.zone.local(1989, "dec", 13, 20, 15, 1)
fetched_at_format = "%D %l:%M%P %Z %z"
efolder_fetched_at_format = "%FT%T.%LZ"
fetched_doc_struct={
      documents: documents,
      manifest_vbms_fetched_at: service_manifest_vbms_fetched_at.utc.strftime(efolder_fetched_at_format),
      manifest_vva_fetched_at: service_manifest_vva_fetched_at.utc.strftime(efolder_fetched_at_format)
    };
require 'rspec/mocks/standalone'
expect(EFolderService).to receive(:fetch_documents_for).and_return(fetched_doc_struct)

document_fetcher = DocumentFetcher.new(appeal: Appeal.last, use_efolder: true);
ActiveRecord::Base.logger.level = :warn
query_data = SqlTracker.track do 
  document_fetcher.find_or_create_documents!
end;
query_data.values.pluck(:count, :duration, :sql)
=> [[1, 5.623, "SELECT \"documents\".* FROM \"documents\" WHERE \"documents\".\"file_number\" = xxx AND \"documents\".\"series_id\" IS NULL"],
 [1, 65.57000000000001, "SELECT \"documents\".* FROM \"documents\" WHERE \"documents\".\"vbms_document_id\" IN (xxx)"],
 [1, 28.608, "SELECT \"annotations\".* FROM \"annotations\" WHERE \"annotations\".\"document_id\" IN (xxx)"],
 [1, 29.114, "SELECT \"documents_tags\".* FROM \"documents_tags\" WHERE \"documents_tags\".\"document_id\" IN (xxx)"],
 [250, 840.9330000000007, "INSERT INTO \"documents\" (\"created_at\", \"file_number\", \"received_at\", \"series_id\", \"type\", \"updated_at\", \"upload_date\", \"vbms_document_id\") VALUES (xxx) RETURNING \"id\""],
 [250, 8999.695000000002, "SELECT 1 AS one FROM \"documents\" WHERE \"documents\".\"series_id\" = xxx AND \"documents\".\"vbms_document_id\" NOT IN (xxx) LIMIT $501"],
 [199, 654.7490000000004, "UPDATE \"documents\" SET \"type\" = xxx, \"updated_at\" = xxx WHERE \"documents\".\"id\" = xxx"]]

# Clean up for next run
Document.where(id:[last_doc_id+1..Document.last.id]).delete_all
```
Close the Rails console. 
Discard all changes: `git checkout -- .`

Checkout the PR branch, then run the same code and compare the benchmark numbers.
Note the UPDATE is replaced with an INSERT by the [`activerecord-import` gem](https://github.com/zdennis/activerecord-import). 
```ruby
=> [[1, 2.754, "SELECT \"documents\".* FROM \"documents\" WHERE \"documents\".\"file_number\" = xxx AND \"documents\".\"series_id\" IS NULL"],
 [1, 37.568, "SELECT \"documents\".\"vbms_document_id\" FROM \"documents\" WHERE \"documents\".\"vbms_document_id\" IN (xxx)"],
 [1, 3.1, "SELECT pg_get_serial_sequence('documents', 'id')"],
 [1,
  20.262999999999998,
  "INSERT INTO \"documents\" (\"id\",\"category_medical\",\"category_other\",\"category_procedural\",\"created_at\",\"description\",\"file_number\",\"previous_document_version_id\",\"received_at\",\"series_id\",\"type\",\"updated_at\",\"upload_date\",\"vbms_document_id\") VALUES (xxx) DO UPDATE SET \"type\"=EXCLUDED.\"type\",\"received_at\"=EXCLUDED.\"received_at\",\"upload_date\"=EXCLUDED.\"upload_date\",\"vbms_document_id\"=EXCLUDED.\"vbms_document_id\",\"series_id\"=EXCLUDED.\"series_id\",\"updated_at\"=EXCLUDED.\"updated_at\" RETURNING \"id\""],
 [3, 73.40899999999999, "SELECT \"documents\".* FROM \"documents\" WHERE \"documents\".\"vbms_document_id\" IN (xxx)"],
 [1,
  15.718,
  "INSERT INTO \"documents\" (\"id\",\"category_medical\",\"category_other\",\"category_procedural\",\"created_at\",\"description\",\"file_number\",\"previous_document_version_id\",\"received_at\",\"series_id\",\"type\",\"updated_at\",\"upload_date\",\"vbms_document_id\") VALUES (xxx) RETURNING \"id\""],
 [1, 21.237, "SELECT \"documents\".\"series_id\" FROM \"documents\" WHERE \"documents\".\"vbms_document_id\" IN (xxx)"],
 [1, 64.21000000000001, "SELECT \"documents\".* FROM \"documents\" WHERE \"documents\".\"series_id\" IN (xxx) AND \"documents\".\"vbms_document_id\" NOT IN (xxx)"]]
```

### Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### Integrations: Adding endpoints for external APIs
* [ ] Integrations impacting functionality are tested in Caseflow UAT

- [ ] For higher-risk changes: [Deploy the custom branch to UAT to test](https://github.com/department-of-veterans-affairs/appeals-deployment/wiki/Applications---Deploy-Custom-Branch-to-UAT)

### CODE REVIEW

**Comment by yoomlam** [comment]
For reference: 3 seconds average response time
image
[link](https://one.newrelic.com/launcher/nr1-core.explorer?platform[accountId]=1788458&platform[timeRange][duration]=1800000&pane=eyJ0cmFuc2FjdGlvbklkIjoiNWIyMjQzNmY2ZTc0NzI2ZjZjNmM2NTcyMmY3MjY1NjE2NDY1NzIyZjY0NmY2Mzc1NmQ2NTZlNzQ3MzJmNjk2ZTY0NjU3ODIyMmMyMjIyNWQiLCJ0eXBlIjoiYXBwIiwibmVyZGxldElkIjoiYXBtLW5lcmRsZXRzLnRyYW5zYWN0aW9ucyIsImVudGl0eUlkIjoiTVRjNE9EUTFPSHhCVUUxOFFWQlFURWxEUVZSSlQwNThPREEyTURnMU5EUSJ9&sidebars[0]=eyJuZXJkbGV0SWQiOiJucjEtY29yZS5hY3Rpb25zIiwiZW50aXR5SWQiOiJNVGM0T0RRMU9IeEJVRTE4UVZCUVRFbERRVlJKVDA1OE9EQTJNRGcxTkRRIiwic2VsZWN0ZWROZXJkbGV0Ijp7Im5lcmRsZXRJZCI6ImFwbS1uZXJkbGV0cy50cmFuc2FjdGlvbnMifX0)

File: spec/services/document_fetcher_spec.rb [hunk 1]
```diff
@@ -314,8 +315,8 @@
           # Uncomment the following to see a count of SQL queries
           # pp query_data.values.pluck(:sql, :count)
           doc_insert_queries = query_data.values.select { |o| o[:sql].start_with?("INSERT INTO \"documents\"") }
-          # This expected count of 25 is bad and will be fixed in another PR
-          expect(doc_insert_queries.pluck(:count).max).to eq 25
+          expect(doc_insert_queries.pluck(:count).max).to eq 1
+          expect(query_data.values.select { |o| o[:sql].start_with?("UPDATE") }).to be_empty
```
Comment (yoomlam) [inline spec/services/document_fetcher_spec.rb:319]: Fixed in this PR!

**Review by ajspotts — APPROVED** [review]
Reader docs load properly ✅ 
Testing plan (comparable loading times) ✅ 
Awesome changes, would be awesome to limit those NewRelic alerts!

**Review by ajspotts — APPROVED** [review]
lgtm!

**Pull request merged**